jpsonic-main/src/main/java/com/tesshu/jpsonic/service/search/IndexManager.java

```diff
@@ -741,23 +739,23 @@ private List<GenreFreq> getGenreFreqs() {
         }
     }
 
-    private int getAlbumGenreCount(IndexSearcher searcher, String genreName, GenreMasterCriteria criteria)
-            throws IOException {
-        if (criteria.scope() != Scope.ALBUM) {
-            return Genre.COUNT_UNACQUIRED;
+    private int getAlbumGenreCount(IndexSearcher searcher, String genreName, GenreMasterCriteria criteria) {
+        try {
+            return searcher.search(queryFactory.getAlbumId3GenreCount(genreName, criteria.folders()),
+                    new TotalHitCountCollectorManager());
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
         }
-        Query query = queryFactory.getAlbumId3GenreCount(genreName, criteria.folders());
-        return searcher.search(query, new TotalHitCountCollectorManager());
     }
 
-    private int getSongGenreCount(IndexSearcher searcher, String genreName, GenreMasterCriteria criteria)
-            throws IOException {
-        if (criteria.scope() == Scope.SONG || criteria.sort() == Sort.SONG_COUNT) {
-            MediaType[] types = criteria.scope() == Scope.SONG ? criteria.types() : MUSIC_AND_AUDIOBOOK;
-            Query query = queryFactory.getSongGenreCount(genreName, criteria.folders(), types);
-            return searcher.search(query, new TotalHitCountCollectorManager());
+    private int getSongGenreCount(IndexSearcher searcher, String genreName, GenreMasterCriteria criteria) {
+        try {
+            MediaType[] types = criteria.types().length == 0 ? MUSIC_AND_AUDIOBOOK : criteria.types();
+            return searcher.search(queryFactory.getSongGenreCount(genreName, criteria.folders(), types),
+                    new TotalHitCountCollectorManager());
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
         }
```

Comment:
To be precise, this fix increases the genre logic according to criteria. However, it is not necessary to reproduce this exactly in your coding. This logic was written to avoid unnecessary counting. However, the counting speed was much faster than expected (fast enough that the extra avoidance was not necessary). So in this commit, they are rather simplified.

Just to be sure, Check the speed using data of 100,000 songs and 100 genres or more.

Reply:
Done.